tests/test_scalars.py

```diff
@@ -0,0 +1,234 @@
+
+
+def test_periof_add_subtract() -> None:
+    p = pd.Period("2012-1-1", freq="D")
+
+    as0 = pd.Timedelta(1, "D")
+    as1 = dt.timedelta(days=1)
+    as2 = np.timedelta64(1, "D")
+    as3 = np.int64(1)
+    as4 = int(1)
+    as5 = pd.period_range("2012-1-1", periods=10, freq="D")
+    as6 = pd.Period("2012-1-1", freq="D")
+    scale = 24 * 60 * 60 * 10**9
+    as7 = cast(TimedeltaSeries, pd.Series(pd.timedelta_range(scale, scale, freq="D")))
+    as8 = pd.Series(as5)
+    as9 = pd.timedelta_range(scale, scale, freq="D")
+    as10 = pd.NaT
+
+    check(assert_type(p + as0, pd.Period), pd.Period)
+    check(assert_type(p + as1, pd.Period), pd.Period)
+    check(assert_type(p + as2, pd.Period), pd.Period)
+    check(assert_type(p + as3, pd.Period), pd.Period)
+    check(assert_type(p + as4, pd.Period), pd.Period)
+    check(assert_type(p + p.freq, pd.Period), pd.Period)
+    check(assert_type(p + (p - as5), pd.PeriodIndex), pd.PeriodIndex)
+    check(assert_type(p + as7, PeriodSeries), pd.Series, pd.Period)
+    check(assert_type(p + as9, pd.PeriodIndex), pd.PeriodIndex)
+    check(assert_type(p + as10, NaTType), NaTType)
+    das8 = cast(TimedeltaSeries, (as8 - as8))
+    check(assert_type(p + das8, PeriodSeries), pd.Series, pd.Period)
+    check(assert_type(p - as0, pd.Period), pd.Period)
+    check(assert_type(p - as1, pd.Period), pd.Period)
+    check(assert_type(p - as2, pd.Period), pd.Period)
+    check(assert_type(p - as3, pd.Period), pd.Period)
+    check(assert_type(p - as4, pd.Period), pd.Period)
+    check(assert_type(p - as5, pd.Index), pd.Index)
+    check(assert_type(p - as6, BaseOffset), Day)
+    check(assert_type(p - as7, PeriodSeries), pd.Series, pd.Period)
+    check(assert_type(p - as9, pd.PeriodIndex), pd.PeriodIndex)
+    check(assert_type(p - as10, NaTType), NaTType)
+    check(assert_type(p - p.freq, pd.Period), pd.Period)
+
+    check(assert_type(as0 + p, pd.Period), pd.Period)
+    check(assert_type(as1 + p, pd.Period), pd.Period)
+    check(assert_type(as2 + p, pd.Period), pd.Period)
+    check(assert_type(as3 + p, pd.Period), pd.Period)
+    check(assert_type(as4 + p, pd.Period), pd.Period)
+    check(assert_type(as7 + p, PeriodSeries), pd.Series, pd.Period)
+    # TODO: Improve Index to not handle __add__(period)
+    check(assert_type(as9 + p, pd.Index), pd.PeriodIndex)
+    check(assert_type(as10 + p, NaTType), NaTType)
+    check(assert_type(p.freq + p, pd.Period), pd.Period)
+
+    check(assert_type(as5 - p, pd.Index), pd.Index)
+
+    check(assert_type(p.__radd__(as0), pd.Period), pd.Period)
+    check(assert_type(p.__radd__(as1), pd.Period), pd.Period)
+    check(assert_type(p.__radd__(as2), pd.Period), pd.Period)
+    check(assert_type(p.__radd__(as3), pd.Period), pd.Period)
+    check(assert_type(p.__radd__(as4), pd.Period), pd.Period)
+    check(assert_type(p.__radd__(as10), NaTType), NaTType)
+    check(assert_type(p.__radd__(p.freq), pd.Period), pd.Period)
```

Comment:
let's not call `__radd__()` directly and just do things like `p.freq + p`, which should cause the type checker to look at `__radd__()`

Reply:
Yes, I agree that's true for debugging purposes.  If you want, you could do it as `p.freq + p` and add a comment that you are testing `p.__radd__(p.freq)` .  The reason I want to use the `p.freq +  p` expression is that it directly tests how the rules for `__add__()` and `__radd__()` work.

Reply:
Added some notes and reorganized these.  Reog make is clear that operator comes with each call to `__radd/sub__`